File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    private boolean graphDFSIve(int currVertex, Map<Integer, List<Integer>> adjList, int destination){
        Map<Integer, Boolean> visitedVerticesMap = new HashMap<>();

        Stack<Integer> verticesStack = new Stack<>();

        verticesStack.push(currVertex);
        visitedVerticesMap.put(currVertex, true);

        boolean isPathExist = false;

        while(!verticesStack.isEmpty()){
            currVertex = verticesStack.pop();

            if(currVertex == destination){
                isPathExist = true;
                break;
            }

            List<Integer> currVertexNeighbors = adjList.get(currVertex);
            for(int idxI = 0; idxI < currVertexNeighbors.size(); idxI++){
                int currNeighborVertex = currVertexNeighbors.get(idxI);

                if(!visitedVerticesMap.containsKey(currNeighborVertex)){
                    visitedVerticesMap.put(currNeighborVertex, true);
                    verticesStack.push(currNeighborVertex);
                }
            }
        }

        return isPathExist;
    }

    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> adjList = new HashMap<>();

        for(int[] currEdge : edges){
            int vertex1 = currEdge[0];
            int vertex2 = currEdge[1];

            if(!adjList.containsKey(vertex1)){
                adjList.put(vertex1, new ArrayList<>());
            }   
            if(!adjList.containsKey(vertex2)){
                adjList.put(vertex2, new ArrayList<>());
            }

            adjList.get(vertex1).add(vertex2);
            adjList.get(vertex2).add(vertex1);
        }
        

        boolean isPathExists = graphDFSIve(source, adjList, destination);
        return isPathExists;
    }
}

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        # Logic: Simple BFS/DFS can do the job
        # start from the source vertex and then 
        # if you ever reach the destination vertex
        # that means path exists else path does not exist;

        adj_list = {}
        visited_vertices_map = {}
        is_path_exist = False

        for edge in edges:
            vertex1 = edge[0]
            vertex2 = edge[1]

            if vertex1 not in adj_list:
                adj_list[vertex1] = []

            if vertex2 not in adj_list:
                adj_list[vertex2] = []

            adj_list[vertex1].append(vertex2)
            adj_list[vertex2].append(vertex1)

        def graphDFSIve(curr_vertex):
            vertices_stack = []

            vertices_stack.append(curr_vertex)
            is_path_exist = False

            while len(vertices_stack) > 0:
                curr_vertex = vertices_stack.pop()

                if curr_vertex == destination:
                    is_path_exist = True
                    break

                if curr_vertex not in visited_vertices_map:
                    visited_vertices_map[curr_vertex] = 'visited'
                    curr_vertex_neighbors = adj_list[curr_vertex]

                    for curr_neighbor_vertex in curr_vertex_neighbors:
                        if curr_neighbor_vertex not in visited_vertices_map:
                            vertices_stack.append(curr_neighbor_vertex)
            
            return is_path_exist
        
        is_path_exist = graphDFSIve(source)

        return is_path_exist